File: src/services/modelRegistry.ts
import { supabase } from '@/lib/supabase';

export interface ModelRegistryRow {
  id: string;
  model_name: string;
  version: string;
  description: string | null;
  artifact_path: string | null;
  created_by: string | null;
  created_at: string;
  metadata: Record<string, unknown> | null;
}

export interface ModelEvaluationRow {
  id: string;
  model_id: string;
  accuracy: number;
  top3_accuracy: number;
  confusion: Record<string, Record<string, number>>;
  report_path: string | null;
  created_at: string;
}

export async function registerModelVersion(payload: {
  modelName: string;
  version: string;
  description?: string;
  artifactPath?: string;
  metadata?: Record<string, unknown>;
}): Promise<ModelRegistryRow> {
  const { data, error } = await supabase
    .from('model_registry')
    .insert({
      model_name: payload.modelName,
      version: payload.version,
      description: payload.description ?? null,
      artifact_path: payload.artifactPath ?? null,
      metadata: payload.metadata ?? {},
    })
    .select()
    .single();

  if (error) {
    throw error;
  }

  return data as ModelRegistryRow;
}

export async function recordModelEvaluation(payload: {
  modelId: string;
  accuracy: number;
  top3Accuracy: number;
  confusion: Record<string, Record<string, number>>;
  reportPath?: string;
}): Promise<ModelEvaluationRow> {
  const { data, error } = await supabase
    .from('model_evaluations')
    .insert({
      model_id: payload.modelId,
      accuracy: payload.accuracy,
      top3_accuracy: payload.top3Accuracy,
      confusion: payload.confusion,
      report_path: payload.reportPath ?? null,
    })
    .select()
    .single();

  if (error) {
    throw error;
  }

  return data as ModelEvaluationRow;
}

export async function fetchLatestModel(modelName: string): Promise<ModelRegistryRow | null> {
  const { data, error } = await supabase
    .from('model_registry')
    .select('*')
    .eq('model_name', modelName)
    .order('created_at', { ascending: false })
    .limit(1)
    .maybeSingle();

  if (error) {
    throw error;
  }

  return (data as ModelRegistryRow | null) ?? null;
}

export async function listModelEvaluations(modelId: string): Promise<ModelEvaluationRow[]> {
  const { data, error } = await supabase
    .from('model_evaluations')
    .select('*')
    .eq('model_id', modelId)
    .order('created_at', { ascending: false });

  if (error) {
    throw error;
  }

  return (data ?? []) as ModelEvaluationRow[];
}

File: services/python-ai/app/orchestrator/model_registry.py
from typing import Any, Dict, List, Optional

from ..agent.utils.nanoid import nanoid

_models: List[Dict[str, Any]] = []
_evaluations: List[Dict[str, Any]] = []


def register_model_version(payload: Dict[str, Any]) -> Dict[str, Any]:
    row = {
        "id": nanoid(),
        "model_name": payload.get("modelName"),
        "version": payload.get("version"),
        "description": payload.get("description"),
        "artifact_path": payload.get("artifactPath"),
        "created_by": payload.get("created_by"),
        "created_at": "",
        "metadata": payload.get("metadata") or None,
    }
    _models.append(row)
    return row


def record_model_evaluation(payload: Dict[str, Any]) -> Dict[str, Any]:
    row = {
        "id": nanoid(),
        "model_id": payload.get("modelId"),
        "accuracy": payload.get("accuracy"),
        "top3_accuracy": payload.get("top3Accuracy"),
        "confusion": payload.get("confusion") or {},
        "report_path": payload.get("reportPath"),
        "created_at": "",
    }
    _evaluations.append(row)
    return row


def fetch_latest_model(model_name: str) -> Optional[Dict[str, Any]]:
    candidates = [row for row in _models if row.get("model_name") == model_name]
    return candidates[-1] if candidates else None


def list_model_evaluations(model_id: str) -> List[Dict[str, Any]]:
    return [row for row in _evaluations if row.get("model_id") == model_id]

